From 1096088d78f9df88aaeacde68e0decde70b48f6e Mon Sep 17 00:00:00 2001
From: Andrei Warkentin <andrei.warkentin@intel.com>
Date: Thu, 16 Mar 2023 15:46:04 -0500
Subject: [PATCH 1/1] RiscVVirt: Patches to enable FdtBusPkg components

Ports over the following components to use FdtBusDxe
and driver binding where possible.

- VirtioFdtDxe
- PciHostBridgeLibEcam
- FdtPciPcdProducerLib
- HighMemDxe (binding variant by default)
- PciSioSerialDxe (which supports the FDT UART device),
  instead of SerialDxe.

PlatformBootManagerLib is modified to set ConOut/ConIn/StdErr
to the UART described by /chosen/stdout-path.

Signed-off-by: Andrei Warkentin <andrei.warkentin@intel.com>
---
 OvmfPkg/RiscVVirt/RiscVVirtQemu.dsc           | 26 ++++--
 OvmfPkg/RiscVVirt/RiscVVirtQemu.fdf           | 11 ++-
 .../PlatformBootManagerLib.inf                |  4 +
 .../PlatformBootManagerLib/PlatformBm.c       | 92 ++++++++++++-------
 OvmfPkg/RiscVVirt/Sec/Platform.c              |  1 -
 5 files changed, 91 insertions(+), 43 deletions(-)

diff --git a/OvmfPkg/RiscVVirt/RiscVVirtQemu.dsc b/OvmfPkg/RiscVVirt/RiscVVirtQemu.dsc
index f8b9479345d7..122b7464846a 100644
--- a/OvmfPkg/RiscVVirt/RiscVVirtQemu.dsc
+++ b/OvmfPkg/RiscVVirt/RiscVVirtQemu.dsc
@@ -95,13 +95,15 @@ [LibraryClasses.common]
   FrameBufferBltLib|MdeModulePkg/Library/FrameBufferBltLib/FrameBufferBltLib.inf
   QemuBootOrderLib|OvmfPkg/Library/QemuBootOrderLib/QemuBootOrderLib.inf
   FileExplorerLib|MdeModulePkg/Library/FileExplorerLib/FileExplorerLib.inf
-  PciPcdProducerLib|OvmfPkg/Fdt/FdtPciPcdProducerLib/FdtPciPcdProducerLib.inf
+  # PciPcdProducerLib|OvmfPkg/Fdt/FdtPciPcdProducerLib/FdtPciPcdProducerLib.inf
+  PciPcdProducerLib|FdtBusPkg/Library/FdtPciPcdProducerLib/FdtPciPcdProducerLib.inf
   PciSegmentLib|MdePkg/Library/BasePciSegmentLibPci/BasePciSegmentLibPci.inf
-  PciHostBridgeLib|OvmfPkg/Fdt/FdtPciHostBridgeLib/FdtPciHostBridgeLib.inf
+  PciHostBridgeLib|FdtBusPkg/Library/PciHostBridgeLibEcam/PciHostBridgeLibEcam.inf
   PciHostBridgeUtilityLib|OvmfPkg/Library/PciHostBridgeUtilityLib/PciHostBridgeUtilityLib.inf
   PeiHardwareInfoLib|OvmfPkg/Library/HardwareInfoLib/PeiHardwareInfoLib.inf
   PlatformHookLib|MdeModulePkg/Library/BasePlatformHookLibNull/BasePlatformHookLibNull.inf
   ImagePropertiesRecordLib|MdeModulePkg/Library/ImagePropertiesRecordLib/ImagePropertiesRecordLib.inf
+  FbpUtilsLib|FdtBusPkg/Library/FbpUtilsLib/FbpUtilsLib.inf
 
 !if $(TPM2_ENABLE) == TRUE
   Tpm2CommandLib|SecurityPkg/Library/Tpm2CommandLib/Tpm2CommandLib.inf
@@ -334,7 +336,8 @@ [Components]
   MdeModulePkg/Universal/Console/ConSplitterDxe/ConSplitterDxe.inf
   MdeModulePkg/Universal/Console/GraphicsConsoleDxe/GraphicsConsoleDxe.inf
   MdeModulePkg/Universal/Console/TerminalDxe/TerminalDxe.inf
-  MdeModulePkg/Universal/SerialDxe/SerialDxe.inf
+  # MdeModulePkg/Universal/SerialDxe/SerialDxe.inf
+  FdtBusPkg/Drivers/PciSioSerialDxe/PciSioSerialDxe.inf
 
   MdeModulePkg/Universal/HiiDatabaseDxe/HiiDatabaseDxe.inf
 
@@ -351,14 +354,18 @@ [Components]
   #
   # Platform Driver
   #
-  OvmfPkg/Fdt/VirtioFdtDxe/VirtioFdtDxe.inf
+  # OvmfPkg/Fdt/VirtioFdtDxe/VirtioFdtDxe.inf
+  FdtBusPkg/Drivers/VirtioFdtDxe/VirtioFdtDxe.inf
+  FdtBusPkg/Drivers/FdtBusDxe/FdtBusDxe.inf
   EmbeddedPkg/Drivers/FdtClientDxe/FdtClientDxe.inf {
     <LibraryClasses>
       DevicePathLib|MdePkg/Library/UefiDevicePathLib/UefiDevicePathLib.inf
       PcdLib|MdePkg/Library/BasePcdLibNull/BasePcdLibNull.inf
       DebugLib|MdePkg/Library/BaseDebugLibNull/BaseDebugLibNull.inf
   }
-  OvmfPkg/Fdt/HighMemDxe/HighMemDxe.inf
+  #OvmfPkg/Fdt/HighMemDxe/HighMemDxe.inf
+  FdtBusPkg/Drivers/HighMemDxe/HighMemDxe.inf
+  #FdtBusPkg/Drivers/HighMemDxe/HighMemDxeNoBinding.inf
   OvmfPkg/VirtioBlkDxe/VirtioBlk.inf
   OvmfPkg/VirtioScsiDxe/VirtioScsi.inf
   OvmfPkg/VirtioNetDxe/VirtioNet.inf
@@ -448,12 +455,14 @@ [Components]
   #
   OvmfPkg/RiscVVirt/PciCpuIo2Dxe/PciCpuIo2Dxe.inf {
     <LibraryClasses>
-      NULL|OvmfPkg/Fdt/FdtPciPcdProducerLib/FdtPciPcdProducerLib.inf
+      # NULL|OvmfPkg/Fdt/FdtPciPcdProducerLib/FdtPciPcdProducerLib.inf
+      NULL|FdtBusPkg/Library/FdtPciPcdProducerLib/FdtPciPcdProducerLib.inf
   }
   MdeModulePkg/Bus/Pci/PciHostBridgeDxe/PciHostBridgeDxe.inf
   MdeModulePkg/Bus/Pci/PciBusDxe/PciBusDxe.inf {
     <LibraryClasses>
-      NULL|OvmfPkg/Fdt/FdtPciPcdProducerLib/FdtPciPcdProducerLib.inf
+      # NULL|OvmfPkg/Fdt/FdtPciPcdProducerLib/FdtPciPcdProducerLib.inf
+      NULL|FdtBusPkg/Library/FdtPciPcdProducerLib/FdtPciPcdProducerLib.inf
   }
   OvmfPkg/PciHotPlugInitDxe/PciHotPlugInit.inf
   OvmfPkg/VirtioPciDeviceDxe/VirtioPciDeviceDxe.inf
@@ -503,5 +512,6 @@ [Components]
   MdeModulePkg/Universal/Acpi/BootGraphicsResourceTableDxe/BootGraphicsResourceTableDxe.inf
   OvmfPkg/AcpiPlatformDxe/AcpiPlatformDxe.inf {
     <LibraryClasses>
-      NULL|OvmfPkg/Fdt/FdtPciPcdProducerLib/FdtPciPcdProducerLib.inf
+      # NULL|OvmfPkg/Fdt/FdtPciPcdProducerLib/FdtPciPcdProducerLib.inf
+      NULL|FdtBusPkg/Library/FdtPciPcdProducerLib/FdtPciPcdProducerLib.inf
   }
diff --git a/OvmfPkg/RiscVVirt/RiscVVirtQemu.fdf b/OvmfPkg/RiscVVirt/RiscVVirtQemu.fdf
index 40d12e0f4c46..792144bfcab6 100644
--- a/OvmfPkg/RiscVVirt/RiscVVirtQemu.fdf
+++ b/OvmfPkg/RiscVVirt/RiscVVirtQemu.fdf
@@ -69,9 +69,13 @@ [FV.DXEFV]
 INF  MdeModulePkg/Core/Dxe/DxeMain.inf
 INF  MdeModulePkg/Universal/PCD/Dxe/Pcd.inf
 INF  MdeModulePkg/Universal/DevicePathDxe/DevicePathDxe.inf
-INF  OvmfPkg/Fdt/VirtioFdtDxe/VirtioFdtDxe.inf
+# INF  OvmfPkg/Fdt/VirtioFdtDxe/VirtioFdtDxe.inf
+INF  FdtBusPkg/Drivers/VirtioFdtDxe/VirtioFdtDxe.inf
+INF  FdtBusPkg/Drivers/FdtBusDxe/FdtBusDxe.inf
 INF  EmbeddedPkg/Drivers/FdtClientDxe/FdtClientDxe.inf
-INF  OvmfPkg/Fdt/HighMemDxe/HighMemDxe.inf
+# INF  OvmfPkg/Fdt/HighMemDxe/HighMemDxe.inf
+INF  FdtBusPkg/Drivers/HighMemDxe/HighMemDxe.inf
+# INF  FdtBusPkg/Drivers/HighMemDxe/HighMemDxeNoBinding.inf
 
 #
 # PI DXE Drivers producing Architectural Protocols (EFI Services)
@@ -98,7 +102,8 @@ [FV.DXEFV]
 INF  MdeModulePkg/Universal/Console/ConSplitterDxe/ConSplitterDxe.inf
 INF  MdeModulePkg/Universal/Console/GraphicsConsoleDxe/GraphicsConsoleDxe.inf
 INF  MdeModulePkg/Universal/Console/TerminalDxe/TerminalDxe.inf
-INF  MdeModulePkg/Universal/SerialDxe/SerialDxe.inf
+# INF  MdeModulePkg/Universal/SerialDxe/SerialDxe.inf
+INF  FdtBusPkg/Drivers/PciSioSerialDxe/PciSioSerialDxe.inf
 
 # RISC-V Core Drivers
 INF  UefiCpuPkg/CpuTimerDxeRiscV64/CpuTimerDxeRiscV64.inf
diff --git a/OvmfPkg/RiscVVirt/Library/PlatformBootManagerLib/PlatformBootManagerLib.inf b/OvmfPkg/RiscVVirt/Library/PlatformBootManagerLib/PlatformBootManagerLib.inf
index 9d66c8110c53..b62c2af15da1 100644
--- a/OvmfPkg/RiscVVirt/Library/PlatformBootManagerLib/PlatformBootManagerLib.inf
+++ b/OvmfPkg/RiscVVirt/Library/PlatformBootManagerLib/PlatformBootManagerLib.inf
@@ -32,6 +32,7 @@ [Packages]
   OvmfPkg/OvmfPkg.dec
   SecurityPkg/SecurityPkg.dec
   ShellPkg/ShellPkg.dec
+  FdtBusPkg/FdtBusPkg.dec
 
 [LibraryClasses]
   BaseLib
@@ -50,6 +51,7 @@ [LibraryClasses]
   UefiBootServicesTableLib
   UefiLib
   UefiRuntimeServicesTableLib
+  FbpUtilsLib
 
 [FixedPcd]
   gEfiMdePkgTokenSpaceGuid.PcdUartDefaultBaudRate
@@ -73,3 +75,5 @@ [Protocols]
   gEfiGraphicsOutputProtocolGuid
   gEfiPciRootBridgeIoProtocolGuid
   gVirtioDeviceProtocolGuid
+  gEfiDtIoProtocolGuid
+  gEfiDevicePathProtocolGuid
diff --git a/OvmfPkg/RiscVVirt/Library/PlatformBootManagerLib/PlatformBm.c b/OvmfPkg/RiscVVirt/Library/PlatformBootManagerLib/PlatformBm.c
index 44fd21f74fcf..3b7157b7a73e 100644
--- a/OvmfPkg/RiscVVirt/Library/PlatformBootManagerLib/PlatformBm.c
+++ b/OvmfPkg/RiscVVirt/Library/PlatformBootManagerLib/PlatformBm.c
@@ -18,6 +18,8 @@
 #include <Library/QemuBootOrderLib.h>
 #include <Library/TpmPlatformHierarchyLib.h>
 #include <Library/UefiBootManagerLib.h>
+#include <Library/DevicePathLib.h>
+#include <Library/FbpUtilsLib.h>
 #include <Protocol/DevicePath.h>
 #include <Protocol/FirmwareVolume2.h>
 #include <Protocol/GraphicsOutput.h>
@@ -25,6 +27,7 @@
 #include <Protocol/PciIo.h>
 #include <Protocol/PciRootBridgeIo.h>
 #include <Protocol/VirtioDevice.h>
+#include <Protocol/DtIo.h>
 #include <Guid/EventGroup.h>
 #include <Guid/GlobalVariable.h>
 #include <Guid/RootBridgesConnectedEventGroup.h>
@@ -39,27 +42,18 @@
 
 #pragma pack (1)
 typedef struct {
-  VENDOR_DEVICE_PATH            SerialDxe;
   UART_DEVICE_PATH              Uart;
   VENDOR_DEFINED_DEVICE_PATH    TermType;
   EFI_DEVICE_PATH_PROTOCOL      End;
 } PLATFORM_SERIAL_CONSOLE;
 #pragma pack ()
 
-STATIC PLATFORM_SERIAL_CONSOLE  mSerialConsole = {
-  //
-  // VENDOR_DEVICE_PATH SerialDxe
-  //
-  {
-    { HARDWARE_DEVICE_PATH,  HW_VENDOR_DP, DP_NODE_LEN (VENDOR_DEVICE_PATH) },
-    EDKII_SERIAL_PORT_LIB_VENDOR_GUID
-  },
-
+STATIC PLATFORM_SERIAL_CONSOLE  mSerialConsoleSuffix = {
   //
   // UART_DEVICE_PATH Uart
   //
   {
-    { MESSAGING_DEVICE_PATH, MSG_UART_DP,  DP_NODE_LEN (UART_DEVICE_PATH)   },
+    { MESSAGING_DEVICE_PATH, MSG_UART_DP, DP_NODE_LEN (UART_DEVICE_PATH) },
     0,                                      // Reserved
     FixedPcdGet64 (PcdUartDefaultBaudRate), // BaudRate
     FixedPcdGet8 (PcdUartDefaultDataBits),  // DataBits
@@ -767,6 +761,12 @@ PlatformRegisterOptionsAndKeys (
 // BDS Platform Functions
 //
 
+VOID
+EFIAPI
+EfiBootManagerConnectAllConsoles (
+  VOID
+  );
+
 /**
   Do the platform init, can be customized by OEM/IBV
   Possible things that can be done in PlatformBootManagerBeforeConsole:
@@ -784,9 +784,14 @@ PlatformBootManagerBeforeConsole (
   VOID
   )
 {
-  UINT16         FrontPageTimeout;
-  RETURN_STATUS  PcdStatus;
-  EFI_STATUS     Status;
+  UINT16                    FrontPageTimeout;
+  RETURN_STATUS             PcdStatus;
+  EFI_STATUS                Status;
+  EFI_HANDLE                Handle;
+  EFI_DT_IO_PROTOCOL        *DtIo;
+  EFI_DEVICE_PATH_PROTOCOL  *DtDp;
+  EFI_DEVICE_PATH_PROTOCOL  *NewDp;
+  CONST CHAR8               *DtStdOutPath;
 
   //
   // Signal EndOfDxe PI Event
@@ -840,23 +845,42 @@ PlatformBootManagerBeforeConsole (
   //
   // Add the hardcoded serial console device path to ConIn, ConOut, ErrOut.
   //
-  CopyGuid (&mSerialConsole.TermType.Guid, &gEfiTtyTermGuid);
-
-  EfiBootManagerUpdateConsoleVariable (
-    ConIn,
-    (EFI_DEVICE_PATH_PROTOCOL *)&mSerialConsole,
-    NULL
-    );
-  EfiBootManagerUpdateConsoleVariable (
-    ConOut,
-    (EFI_DEVICE_PATH_PROTOCOL *)&mSerialConsole,
-    NULL
-    );
-  EfiBootManagerUpdateConsoleVariable (
-    ErrOut,
-    (EFI_DEVICE_PATH_PROTOCOL *)&mSerialConsole,
-    NULL
-    );
+  CopyGuid (&mSerialConsoleSuffix.TermType.Guid, &gEfiTtyTermGuid);
+
+  DtIo = FbpGetDtRoot ();
+  ASSERT (DtIo != NULL);
+
+  Status = DtIo->Lookup (DtIo, "/chosen", TRUE, &Handle);
+  ASSERT_EFI_ERROR (Status);
+
+  Status = gBS->HandleProtocol (
+                  Handle,
+                  &gEfiDtIoProtocolGuid,
+                  (VOID **)&DtIo
+                  );
+  ASSERT_EFI_ERROR (Status);
+
+  DtStdOutPath = NULL;
+  Status       = DtIo->GetString (DtIo, "stdout-path", 0, &DtStdOutPath);
+  ASSERT_EFI_ERROR (Status);
+  ASSERT (DtStdOutPath != NULL);
+
+  Status = DtIo->Lookup (DtIo, DtStdOutPath, TRUE, &Handle);
+  ASSERT_EFI_ERROR (Status);
+
+  Status = gBS->HandleProtocol (
+                  Handle,
+                  &gEfiDevicePathProtocolGuid,
+                  (VOID **)&DtDp
+                  );
+  ASSERT_EFI_ERROR (Status);
+
+  NewDp = AppendDevicePath (DtDp, (VOID *)&mSerialConsoleSuffix);
+
+  EfiBootManagerUpdateConsoleVariable (ConIn, NewDp, NULL);
+  EfiBootManagerUpdateConsoleVariable (ConOut, NewDp, NULL);
+  EfiBootManagerUpdateConsoleVariable (ErrOut, NewDp, NULL);
+  FreePool (NewDp);
 
   //
   // Set the front page timeout from the QEMU configuration.
@@ -954,6 +978,12 @@ PlatformBootManagerAfterConsole (
     // Connect the rest of the devices.
     //
     EfiBootManagerConnectAll ();
+    //
+    // Uncomment if you want to use an alternate serial console
+    // instead of the one specified by /chosen/stdout-path.
+    //
+    // EfiBootManagerConnectAllConsoles ();
+    //
   }
 
   //
diff --git a/OvmfPkg/RiscVVirt/Sec/Platform.c b/OvmfPkg/RiscVVirt/Sec/Platform.c
index c66432473067..2fd84ef2bca4 100644
--- a/OvmfPkg/RiscVVirt/Sec/Platform.c
+++ b/OvmfPkg/RiscVVirt/Sec/Platform.c
@@ -140,7 +140,6 @@ PlatformPeimInitialization (
 
   PopulateIoResources (Base, "ns16550a");
   PopulateIoResources (Base, "qemu,fw-cfg-mmio");
-  PopulateIoResources (Base, "virtio,mmio");
 
   return EFI_SUCCESS;
 }
-- 
2.34.1

